### DISCUSSION TITLE
What IB software is required for ib-api?

### DISCUSSION THREAD

**Discussion by georgeu2000** [ib-ruby/ib-api#20] [General] [answered]
Currently IB provides 3 downloads, TWS, TWS API and IB Gateway. It was not clear to me from the documentation what is required for ib-api. Does it work with all 3 of them?

https://ib-ruby.github.io/ib-doc/connect_to_tws.html

**Comment by topofocus** [comment]
ib-api is the core gem.
It sends messages to the tws and handles the response from the tws.
It provides an IB::Connection-Instance. 
`/bin/console` loads this environment

ib-gateway adds an additional layer and provides a more object orientated access. IB::Connection is still available (through IB::Connection.current or IB::Gateway.current.tws) 
`/bin/gateway` loads this environment  

hope this clarifies it.

**Comment by georgeu2000** [comment]
Hi, 

Yes, that helps. Thank you.

And both ib-api and ib-gateway work with TSW and IB Gateway software? 

I gather that TWS API is separate and not used...

**Reply by topofocus** [reply]
Any currently supported IB-Gateway or IB-TWS  Instance can be used. Take a look to the setup of ib-container. 
Will provide a final update to Version 972 and then upgrade to the recent 10.10 release. Fortunately only a few api-changes  have to be implemented.

**Reply by georgeu2000** [reply]
Great. Thank you.

**Comment by georgeu2000** [comment, marked as answer]
> Any currently supported IB-Gateway or IB-TWS Instance can be used.

Just wanted to provide an "answer".
